syntax = "proto3";

package inference;

// ─── Shared messages ────────────────────────────────────────────────

message SamplingParams {
  float temperature = 1;
  float top_p = 2;
  int32 top_k = 3;
  float repetition_penalty = 4;
}

message LatencyBreakdown {
  float queue_wait_ms = 1;
  float prefill_ms = 2;
  float decode_ms = 3;
  float total_ms = 4;
}

message WorkerMetrics {
  int32 queue_length = 1;
  float arrival_rate = 2;
  float avg_service_time_ms = 3;
  int32 active_requests = 4;
}

message MetricsRequest {}

// ─── Gateway ────────────────────────────────────────────────────────

message InferRequest {
  string request_id = 1;
  string prompt = 2;
  int32 max_tokens = 3;
  SamplingParams sampling_params = 4;
  int32 deadline_ms = 5;
}

message InferResponse {
  string request_id = 1;
  string generated_text = 2;
  int32 tokens_generated = 3;
  LatencyBreakdown latency = 4;
}

message GatewayMetrics {
  int32 prefill_queue_length = 1;
  int32 decode_queue_length = 2;
  float arrival_rate = 3;
  int32 active_requests = 4;
}

service GatewayService {
  rpc Infer (InferRequest) returns (InferResponse);
  rpc GetMetrics (MetricsRequest) returns (GatewayMetrics);
}

// ─── Prefill ────────────────────────────────────────────────────────

message PrefillRequest {
  string request_id = 1;
  repeated int32 token_ids = 2;
  string model_name = 3;
  SamplingParams sampling_params = 4;
}

message PrefillResponse {
  string request_id = 1;
  string kv_cache_key = 2;
  int32 first_token_id = 3;
  float prefill_time_ms = 4;
}

service PrefillService {
  rpc RunPrefill (PrefillRequest) returns (PrefillResponse);
  rpc GetMetrics (MetricsRequest) returns (WorkerMetrics);
}

// ─── Decode ─────────────────────────────────────────────────────────

message DecodeRequest {
  string request_id = 1;
  string kv_cache_key = 2;
  int32 first_token_id = 3;
  int32 max_tokens = 4;
  SamplingParams sampling_params = 5;
}

message DecodeResponse {
  string request_id = 1;
  int32 token_id = 2;
  string token_text = 3;
  bool is_finished = 4;
  int32 tokens_generated = 5;
  float decode_time_ms = 6;
}

service DecodeService {
  rpc RunDecode (DecodeRequest) returns (stream DecodeResponse);
  rpc GetMetrics (MetricsRequest) returns (WorkerMetrics);
}
